### DISCUSSION TITLE
Constraint Logic

### DISCUSSION THREAD

**Discussion by ghost** [h3rald/min#136] [Features: Implemented]
I looked what we have and saw we only have *type union* and *generics*. However we need more logical operations on constraints i think:

- And
- Not
- Or
- Xor
- ...

I thought a way to do: `(...)` -> *constraint-quotes*

For examples:

```
( symbol react
 ( ( (1 2 3) or ) :signal ==> string :reaction)
 (...)
) ::

( symbol int-xor-float
 ( ( (int floatic-int) xor ) :value ==>)
 (value succ puts!)
) ::

( symbol i-want-pi
 ( ( (int pi) or ) :value ==>)
 (value puts!)
) ::

( symbol i-want-only-one
 ( (1) :value ==>)
 (value puts!)
) ::

...
```

**Comment by ghost** [comment]
Did you take a look that? What do you think about it? Not good? or good?

**Reply by h3rald** [reply]
Well... the truth is that everything you are proposing here can currently be implemented with type classes. Furthermore (I think) you can also use type classes in type unions!

I don't like the idea of allowing a full quotation in a type signature to obtain a type, and Boolean logic isn't short-circuited (unless you use `&&` or `||`, and they are more verbose).

The point of a type signature should be making an operator more readable, so perhaps I would consider some shorthand notation similar to type unions, like:

    more-than-two&less-than-five

But that would have to be micro-parsed on its own (like type unions) so I ask you: how frequent are these cases? I can't even think of a proper example with xor! 

This language is getting closer and closer to Haskell every day... 😜

**Reply by ghost** [reply]
> how frequent are these cases? I can't even think of a proper example with xor!

😃 😄 😆  I understand you. Yes there may be rare examples that includes xor.

> The point of a type signature should be making an operator more readable

Maybe but my idea is to extend the current constraint usement state. It's not capable to make all things with itself.

> you can also use type classes in type unions!

Do you really want to that i write all type classes as for disposable logical constraints?? You joke me probably.

> how frequent are these cases?

There is an example who i know that i stuck there(most of my ideas based on errors/bugs/problems):

```assembly
; Two elements needed for comparison
( typeclass compare_quot
    (quot :q ==> bool :r)
    (q size 2 == @r)
) ::

; We force the quot for containing same elems in itself
( typeclass same_quot
    (quot :q ==> bool :r)
    (
        q first type :T
        q (type T ==) all? @r
    )
) ::

; Lambdic operator for 'most'
( symbol big-of
    ( (compare_quot  same_quot   and) :qc ==> a :result)
    ( (qc first qc last >) (qc first @result return) when qc last @result)
) ::

; (4 2 9 3 1)
; start elem -> 4
; take -> 2
; compare 4 & 2 (compare method = big_of(bool) ) -> 4
; continue

( symbol most
    (quot :q 'sym :method ==> a :r)
    (
        q size
        (dup 0 ==) (null @r return) when ; () -> null
        (1 !=) (
            q 0 1 slice  method -> ; (4 2) big_of -> 4
            q rest rest ; (4 2 9 3 1) pop two -> (9 3 1)
            prepend @q ; prepend 4 -> (4 9 3 1)
            ; go on...
        ) while
        q -> @r
    )
) ::
```

> This language is getting closer and closer to Haskell every day... 😜

... don't care ... we're going to top of min mountain (hopefully).

Addition: Haskell is more complicated to use for me.

**Reply by ghost** [reply]
>  I would consider some shorthand notation similar to type unions, like:
> more-than-two&less-than-five

Don't add it!! This would be make a lot of problems with precedence of logic and would more problems occur.

I think mine is enough to handle with all constraint logic problems. It is easy-readable too.

**Reply by h3rald** [reply]
OK so you are saying that having additional operations on types would let you essentially reuse some type classes, right? That is the only reason though: you are saying that you don't want to create a new type classes every time you want to logically combine some type classes.

Your syntax could even be OK but.., where does it stop? Is it going to always be an operation involving two elements? Or is it going to be more?? Because if it's more than it's not that readable. And the main problem is that that quotation would not be interpreted: it would have a very specific format like:

    ((type1 type2) operation)

But what if you want a more complex operation? Maybe involving or, and, and xor all together? 

At that point I would prefer to implement just type intersections and give you the possibility to write:


    type1&type2&type3&...

...and those would be short-circuited as well! 

Actually, if we add type aliases you could even mix them:


    (
      typealias enumerable 
      (quot|str ==> bool)
    ) ::

So you could write:

   enumerable&type2&type3

Precedence? Right well, I would have to process one operand at a time, so:

1. Check if enumerable is satisfied 
   - check if quot 
   - check if str
2. Check if type2 is satisfied
3. Check if type3 is satisfied
   

What would be the problem with this approach??

**Reply by ghost** [reply]
> you are saying that you don't want to create a new type classes every time you want to logically combine some type classes.

Yep! You got.

> But what if you want a more complex operation? Maybe involving or, and, and xor all together?

Yeah ... because of precedence. I don't like the *ugly* precedence in boolean logic.
Complex operation within boolean logic may blocks all future constraint logic problems so there'll be never to face with again.

> Precedence? Right well, I would have to process one operand at a time, so:

I mean the precedence of `& |` . But hey! You could add the *not oper* too. Because all boolean logic can be remade by *and*, *not* gates (they're basement of logic).

*Or* gate would be fundamental, the precedence of logical operators would be:
- 1. `|`
- 2. `&`
- 3. `not`

So i understood you and then we need three operators there would be `& | not`.
Actually you should find a symbol which represents as the *not* operation.

...and there would be values as expectations because sometimes programmer would like to code switch-like/log(log-level) symbols.

Few examples: `type_a|type_b&type_c` `type_a&type_b&type_c` `1|2|3|4|5`

I suggest you to would choose `!` symbol as the *not* logical operator . It was used at so many languages.

Thanks for all! 👍

**Reply by h3rald** [reply]
That's what I meant... with DeMorgan laws you can rewrite all rules without using parenthesis. It will be more tedious but hopefully you won't need very complex rules (and at worst you can still create a type class).

Yeah I was going to go for `!` for not. 

So basically, the algorithm used to parse the *type expressions* (min parses would still consider them as a single symbol, that's why you can't use spaces) would be:

1. Split by &
2. Process each group and stop when any fails:
   1. Split by |
   2. Process each group and stop when any succeeds 
      1. Validate each type. Negate result if ! precedes the type symbol 

As for values... unfortunately I am afraid you will have to create typeclasses for those: numbers would be ok (kind of) but strings would probably cause the min parser interfere and split the type expression incorrectly.

**Reply by ghost** [reply]
> That's what I meant... with DeMorgan laws you can rewrite all rules without using parenthesis. It will be more tedious but hopefully you won't need very complex rules (and at worst you can still create a type class).
> 
> Yeah I was going to go for `!` for not.
> 
> So basically, the algorithm used to parse the _type expressions_ (min parses would still consider them as a single symbol, that's why you can't use spaces) would be:
> 
> 1. Split by &
> 2. Process each group and stop when any fails:
>    
>    1. Split by |
>    2. Process each group and stop when any succeeds
>       
>       1. Validate each type. Negate result if ! precedes the type symbol

The algorithm and... everything seems okay! 👍 

> As for values... unfortunately I am afraid you will have to create typeclasses for those: numbers would be ok (kind of) but strings would probably cause the min parser interfere and split the type expression incorrectly.

Log symbols oftenly use numbers for log the state. So please only implement numbers(or anything proper like numbers) as values(others ... hopely not needed).

**Reply by h3rald** [reply]
Alright. Let's try with integers only 😊👍

**Reply by h3rald** [reply]
OK I implemented & and ! support as well and it seems to work.

I didn't add integers yet, because I am still a bit skeptical about it: once you add integers then you probably want floats, strings etc... I don't want to rewrite the min parser for type expressions. Let's see how useful Boolean logic with types and typeclasses is first!

**Reply by ghost** [reply]
> OK I implemented & and ! support as well and it seems to work.

Thanks for implementing!

> Let's see how useful Boolean logic with types and typeclasses is first!

Yes.
